src/detect.c

```diff
@@ -1084,16 +1084,16 @@ static bool DetectRunTxInspectRule(ThreadVars *tv,
             TRACE_SID_TXS(s->id, tx, "DetectRunInspectRuleHeader() no match");
             return false;
         }
-        if (DetectEnginePktInspectionRun(tv, det_ctx, s, f, p, NULL) == false) {
-            TRACE_SID_TXS(s->id, tx, "DetectEnginePktInspectionRun no match");
-            return false;
-        }
         /* stream mpm and negated mpm sigs can end up here with wrong proto */
         if (!(AppProtoEquals(s->alproto, f->alproto) || s->alproto == ALPROTO_UNKNOWN)) {
             TRACE_SID_TXS(s->id, tx, "alproto mismatch");
             return false;
         }
     }
+    if (DetectEnginePktInspectionRun(tv, det_ctx, s, f, p, NULL) == false) {
```

Comment:
This does not look correct. I did not understand the explanation, though. Maybe some of the checks are in the wrong inspection list?

Reply:
>  I did not understand the explanation, though.

What did you not understand in the sequence of events produced by the packets ?

> Maybe some of the checks are in the wrong inspection list?

Not sure what an inspection list refers to. Is it something like `DETECT_SM_LIST_MATCH` ?
Then it is `DETECT_SM_LIST_MATCH` for `flowbits` and `g_file_match_list_id` for `file magic`

Otherwise, `flowbits` is run by `DetectEnginePktInspectionRun` because it has a `Match` function, and no `AppLayerTxMatch

> This does not look correct

I thought this was just a performance optimization not to run the check every time
I hoped the CI would catch this if it were not correct.
What should crash in this case ?

Reply:
The concern is performance. Running inspection of things per packet instead of once per tx is potentially a huge cost.

The issue here seems to be that the flowbit logic isn't run at the expected moment. Can you look into ways to get it to run it at the right time w/o brute forcing it?